debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractTransactionCachingLogMinerEventProcessor.java

```diff
@@ -0,0 +1,435 @@
+        return jdbcConnection.connection().prepareStatement(getQueryString(),
+                ResultSet.TYPE_FORWARD_ONLY,
+                ResultSet.CONCUR_READ_ONLY,
+                ResultSet.HOLD_CURSORS_OVER_COMMIT);
+    }
+
+    private void removeEventsWithTransaction(T transaction) {
+        // Clear the event queue for the transaction
+        for (int i = 0; i < transaction.getNumberOfEvents(); ++i) {
+            getEventCache().remove(transaction.getEventId(i));
+        }
+        inMemoryPendingTransactionsCache.remove(transaction.getTransactionId());
+    }
+
+    /**
+     * A comparator that guarantees that the sort order applied to event keys is such that
+     * they are treated as numerical values, sorted as numeric values rather than strings
+     * which would allow "100" to come before "9".
+     */
+    private static class EventKeySortComparator implements Comparator<String> {
+
+        public static EventKeySortComparator INSTANCE = new EventKeySortComparator();
+
+        @Override
+        public int compare(String o1, String o2) {
+            if (o1 == null || !o1.contains("-")) {
+                throw new IllegalStateException("Event Key must be in the format of <transaction>-<event>");
+            }
+            if (o2 == null || !o2.contains("-")) {
+                throw new IllegalStateException("Event Key must be in the format of <transaction>-<event>");
+            }
+            final String[] s1 = o1.split("-");
+            final String[] s2 = o2.split("-");
+
+            // Compare transaction ids, these should generally be identical.
+            int result = s1[0].compareTo(s2[0]);
+            if (result == 0) {
+                result = Long.compare(Long.parseLong(s1[1]), Long.parseLong(s2[1]));
+            }
+            return result;
+        }
+    }
+
+    /**
+     * Purge the necessary caches with all entries that occurred prior to the specified change number.
+     * <p>
+     * NOTE: This method is abstract despite the code used by both all implementations being identical.
+     * This is because the method needed {@code entrySet()} is made available on two different concrete
+     * interfaces between the embedded and remote cache implementations, and therefore we need to access
+     * this method from the concrete implementation classes (RemoteCache and Cache) rather than from
+     * the common class used by CacheProvider (BasicCache).
+     *
+     * @param minCacheScn the minimum system change number to keep entries until
+     */
+    protected void purgeCache(Scn minCacheScn) {
+        getProcessedTransactionsCache().removeIf(entry -> Scn.valueOf(entry.getValue()).compareTo(minCacheScn) < 0);
+        getSchemaChangesCache().removeIf(entry -> Scn.valueOf(entry.getKey()).compareTo(minCacheScn) < 0);
+    }
+
+    @Override
+    protected Scn calculateNewStartScn(Scn endScn, Scn maxCommittedScn) throws InterruptedException {
```

Comment:
You can ignore these for now, this is something I am still working through:

// todo: this logic has changed (at least from the heap), we need to verify no regression introduced.
// todo: this matches the infinispan implementation line-by-line.